kubernetes/kubernetes_service.go

```diff
@@ -358,3 +394,22 @@ func (in *IstioClient) getDeployments(namespace string, deploymentsChan chan dep
 	deployments, err := in.k8s.AppsV1beta1().Deployments(namespace).List(emptyListOptions)
 	deploymentsChan <- deploymentsResponse{deployments: deployments, err: err}
 }
+
+func (in *IstioClient) getPods(namespace string, selector map[string]string, podsChan chan podsResponse) {
+	selectorQuery := labels.Set(selector).String()
+	pods, err := in.GetPods(namespace, selectorQuery)
+	podsChan <- podsResponse{pods: pods, err: err}
+}
+
+func (in *IstioClient) getServicesByApp(namespace string, selector map[string]string, serviceChan chan servicesResponse) {
+	var err error
+	var services *v1.ServiceList
+
+	cfg := config.Get()
+	if serviceSelectorName, ok := selector[cfg.AppLabelName]; ok {
+		serviceSelector := labels.Set{cfg.AppLabelName: serviceSelectorName}
+		services, err = in.k8s.CoreV1().Services(namespace).List(meta_v1.ListOptions{LabelSelector: serviceSelector.String()})
```

Comment:
One question here.
Are we fetching the services that have a label at metadata level, right ?
I think that perhaps is better to match those that have a label in the Selector.
```
apiVersion: v1
kind: Service
metadata:
  name: details
  labels:
    app: details
spec:
  ports:
  - port: 9080
    name: http
  selector:
    app: details
---
```
I mean, on this case, "details" Service define labels on metadata and selector.
The selector labels are the real used to fetch pods, meanwhile (for a Service object) the metadata used are not mandatory.
WDYT ?

Reply:
True. And also consider the case where I might define "details" service but route to a specific version.  Thus allowing me to switch to different versions (canary deployment that I want to revert) by simply changing the selector. In that case, my service may not even have version in its metadata, but the selector will - in this case (see below) the details-v2 workload is the only one that should have this details service:

```
apiVersion: v1
kind: Service
metadata:
  name: details
  labels: <--- notice no version label in the service's labels metadata, only "app" is here
    app: details
spec:
  ports:
  - port: 9080
    name: http
  selector:
    app: details
    version: v2 <--- but notice the selector IS picking a specific version
```

Reply:
As discussed with @xeviknal, another option to query this is iterate through services for a namespace and check if the matching expression of Service's selector works with the Deployment labels.